kit/feature/http_proxy.go

```diff
@@ -0,0 +1,65 @@
+package feature
+
+import (
+	"context"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+
+	"go.uber.org/zap"
+)
+
+// HTTPProxy is an HTTP proxy that's guided by a feature flag. If the feature flag
+// presented to it is enabled, it will perform the proxying behavior. Otherwise
+// it will be a no-op.
+type HTTPProxy struct {
+	proxy   *httputil.ReverseProxy
+	logger  *zap.Logger
+	enabler ProxyEnabler
+}
+
+// NewHTTPProxy returns a new Proxy.
+func NewHTTPProxy(dest *url.URL, logger *zap.Logger, enabler ProxyEnabler) *HTTPProxy {
+	return &HTTPProxy{
+		proxy:   newReverseProxy(dest, enabler.Key()),
+		logger:  logger,
+		enabler: enabler,
+	}
+}
+
+// Do performs the proxying. It returns whether or not the request was proxied.
+func (p *HTTPProxy) Do(w http.ResponseWriter, r *http.Request) bool {
+	if p.enabler.Enabled(r.Context()) {
+		p.proxy.ServeHTTP(w, r)
+		return true
+	}
+	return false
+}
+
+const (
+	// headerProxyFlag is the HTTP header for enriching the request and response
+	// with the feature flag key that precipitated the proxying behavior.
+	headerProxyFlag = "X-Proxy-Flag"
```

Comment:
I recently learned "X-" prefixed headers are discouraged as of 2012: https://stackoverflow.com/questions/3561381/custom-http-headers-naming-conventions

That said, the reasoning is that it makes them hard to promote to public standard later, but that doesn't apply in our case so I think it's fine as is.

I really like that you added this header to the response as well.

Reply:
There is some precedent for using `X-<vendor>-<header-name>`.

We have used `X-INFLUXDB-` and `X-INFLUX` in the past, though those are now deprecated in favor of `X-Platform-`:

https://github.com/influxdata/influxdb/blob/67307d804ede329c5f7a72933ec10680635a7deb/kit/transport/http/api.go#L16

I'd recommend adding that to be consistent.

Reply:
There is some precedent for using `X-<vendor>-<header-name>`.

We have used `X-INFLUXDB-` and `X-INFLUX` in the past, though those are now deprecated in favor of `X-Platform-`:

https://github.com/influxdata/influxdb/blob/67307d804ede329c5f7a72933ec10680635a7deb/kit/transport/http/api.go#L16

I'd recommend adding that to be consistent.